feat: Request Debug Header docs for Kong Gateway 3.5 (#6475)
Kong/docs.konghq.com: app/_src/gateway/production/debug-request.md

---

> title: Debug requests

Author: ❗ For reviewers: I did not test the examples because of the short turn around time. I can make a ticket for post-release to test these unless you think testing these is a blocker for docs release.

Reviewer: I think we should definitely test this guide, but we can publish for now and revisit - as long as we don't forget to do it.

Author: Okay, made a Jira ticket for the week of the 13th to test this

content_type: explanation
---

You can debug requests in {{site.base_gateway}}. This allows an admin user to collect timing information about a given request on demand. Request debugging is triggered using a secure token and the resulting data is returned in a response header named `X-Kong-Request-Debug-Output`. 

Request debugging provides the following insights:
* Time spent in various {{site.base_gateway}} components, such as plugins, DNS resolution, and load balancing. 
* Contextual information, such as the domain name tried during these processes.

{:.note}
> **Note:** This feature is meant for live debugging. The JSON schema of the header containing the timing should never be considered static and is always subject to change.

## Enable request debugging

Request debugging is enabled by default and has the following configurations in [`kong.conf`](/gateway/{{page.kong_version}}/reference/configuration/):

```

> request_debug = on # request debugging is enabled
> request_debug = off # request debugging is disabled

Reviewer: ```suggestion
request_debug = on | off # enable or disable request debugging
```

request_debug_token <token> # Set debug token explicitly. Otherwise, it will be generated randomly when Kong starts, restarts, and reloads. 
```

The debug token (`X-Kong-Request-Debug-Token`) is used to authenticate the client and send the debug request, preventing abuse. 

You can find the debug token in the following locations:
* **{{site.base_gateway}} error log:** The debug token is logged in the error log (notice level) when {{site.base_gateway}} starts, restarts, or reloads. The log line will have the `[request-debug]` prefix to aid in searching.

> * **Filesystem:** The debug token will also be stored in a file at `{prefix}/.request_debug_token` and updated when {{site.base_gateway}} starts, restarts, or reloads.

Reviewer: ```suggestion
* **Filesystem:** The debug token is also stored in a file at `{prefix}/.request_debug_token` and updated when {{site.base_gateway}} starts, restarts, or reloads.
```

## Debug request configuration

> To debug a request, certain request headers must be added. At a minimum, you should set the `X-Kong-Request-Debug` header. If the requests originate from loopback addresses, the `X-Kong-Request-Debug-Token` header also needs to be set.

Reviewer: ```suggestion
To debug a request, add the following request headers:
* At a minimum, you should set the `X-Kong-Request-Debug` header. 
* If the requests originate from loopback addresses, the `X-Kong-Request-Debug-Token` header also needs to be set.
```

                }
            },
            "total_time": 1
        },
        "header_filter": {
            "total_time": 0
        },
        "balancer": {
            "total_time": 0
        },
        "upstream": {
            "total_time": 100,
            "child": {
                "time_to_first_byte": {
                    "total_time": 20
                },
                "streaming": {
                    "Total_time": 80
                }
            }
        }
    }
}
```

You can analyze the debug output like the following:
* The unit of `total_time` is `millisecond`.
* The DNS resolution for `example.host` was cached, which is why it's so fast in the example.
* Upstream took 100ms in this request.
    * The elapsed time from {{site.base_gateway}} sending the request to the upstream to {{site.base_gateway}} receiving the first byte is 20ms.
    * The elapsed time from the first byte to the last byte from the upstream is 80ms.

You can also filter the debug output:

```bash
curl http://kong_proxy_host:8000/example \
    -H "X-Kong-Request-Debug: upstream" \
    -H "X-Kong-Request-Debug-Token: xxxxxx"
```

This will return something like the following:

```bash

> X-Kong-Request-Id: a1a1530f8ddb6f6f2462916ae002b715

Author: ❗ For reviewers: The example output for this example and the following ones didn't match the previous one in the draft doc. I left them as-is just to get this out the door quickly. Should we make them match the first one or is this okay?

X-Kong-Request-Debug-Output: 
{"child":{"upstream":{"total_time":363,"child":{"time_to_first_byte":{"total_time":363}}}},"request_id":"a1a1530f8ddb6f6f2462916ae
002b715"}
```

### Truncation for large debugging output

The downstream system may impose a size restriction on response headers, leading {{site.base_gateway}} to truncate the `X-Kong-Request-Output` if it exceeds 2KB. This truncated output will be unconditionally logged in the `error_log`.

```bash

> curl http://kong_proxy_host:8000/large_debugging_output \

Reviewer: ```suggestion
curl http://localhost:8000/large_debugging_output \
```
